### ISSUE TITLE
Gradient calculation not on ordered sequences

### ISSUE THREAD

**Issue by MengjieZhao** [KBaja/UnsupervisedHI#2]
```
class NG_constraint(tf.keras.layers.Layer):   # Inheritance class
    def __init__(self,alpha,**kwargs):  
        super().__init__(**kwargs)
        self.alpha = alpha
        
    def call(self, inputdata):
        x,t = inputdata
        x_diff = x[1:]-x[:-1]
        t_diff = t[1:]-t[:-1]
        grad = x_diff/t_diff
        loss = tf.reduce_sum(tf.math.maximum(tf.constant(0.0),grad))
        self.add_loss(self.alpha*loss)
        self.add_metric(loss, aggregation='mean', name='Grad')

        return x
    
```

First, thank you for sharing the code of the inspiring work on health index construction!

While I was trying to reimplement the code in PyTorch, I noticed that the non-negative gradient constraint does not take into account that the training data is shuffled. 

My understanding is that in practice, the constraint is not dZ/dt<0, but rather enforcing later sample → lower Z. (Although here one also doesn't distinguish between units, which could be another source of problems).

Please let me know if I have misunderstood the implementation. Either way, I appreciate the inspiration your work has provided.

**Comment by KBaja** [comment]
Hello,

You're absolutely right in your observations. In the original version of the code, shuffling had to be disabled in order to enforce the constraint, which was a limitation on the performance. 

Since then, I’ve made some improvements to better handle this scenario. In short, it’s now possible to ensure that each batch contains observations from the same unit, while still shuffling the data within each unit. The data can then be reordered as needed when applying the constraint later on.

Below, I’ve included a short snippet of the updated PyTorch data loader, along with a revised version of how the negative constraint should be implemented. A similar approach can easily be adapted for TensorFlow as well. 

```
from torch.utils.data import DataLoader, TensorDataset, Sampler

class UnitBatchSampler(Sampler):
    def __init__(self, U_train, batch_size):
        self.U_train = U_train
        self.batch_size = batch_size
        self.unit_indices = {unit: torch.where(U_train == unit)[0].tolist() for unit in torch.unique(U_train)}

    def __iter__(self):
        all_batches = []

        # Create batches for each unit
        for unit, indices in self.unit_indices.items():
            np.random.shuffle(indices)  # Shuffle within each unit
            # unit_batches = [indices[i:i + self.batch_size] 
                            # for i in range(0, len(indices) - len(indices) % self.batch_size, self.batch_size)]
            unit_batches = [indices[i:i + self.batch_size] 
                            for i in range(0, len(indices), self.batch_size)]    
            all_batches.extend(unit_batches)  # Add all unit batches to the list

        # Shuffle all batches across units
        np.random.shuffle(all_batches)

        # Yield batches in a randomized order
        for batch in all_batches:
            yield batch
            
    def __len__(self):
        # Total number of batches across all units
        # return sum(len(indices) // self.batch_size for indices in self.unit_indices.values())
        return sum((len(indices) + self.batch_size - 1) // self.batch_size for indices in self.unit_indices.values())
   
def custom_collate(batch):
    collated_batch = default_collate(batch)
    # Remove batch dimension if batch size is 1
    collated_batch = [x.squeeze(0) if x.shape[0] == 1 else x for x in collated_batch]
    return collated_batch
```

Usage: 
```
    train_tensors = prepare_data(train_data)
    train_ds = TensorDataset(*train_tensors)  # Unpack tensors
    train_ds = DataLoader(train_ds, sampler=UnitBatchSampler(train_data['U'], config['batch_size']),collate_fn=custom_collate)
```

The negative constraint should be rewritten as  :
```
class Constraint(nn.Module):
    def __init__(self, config):
        super(Constraint, self).__init__()
        self.config = config
        self.ng_weight= self.config.get('ng_weight', 1.0)
    def forward(self, x, t):
        sorted_indices = torch.argsort(t, dim=0)
        x_sorted = torch.gather(x, 0, sorted_indices)
        t_sorted = torch.gather(t, 0, sorted_indices)
        
        x_diff = x_sorted[1:] - x_sorted[:-1]
        t_diff = t_sorted[1:] - t_sorted[:-1]
        grad = x_diff/t_diff
        loss = torch.mean(torch.relu(grad))
        # loss = tf.reduce_sum(tf.math.maximum(tf.constant(0.0),grad))
        return loss * self.ng_weight
```

Final Note:
In the negative constraint, as long as the data is properly ordered, it's not necessary to compute the derivative with respect to time (`dZ/dt`). Simply evaluating the change in `Z (dZ)` is sufficient.
